### PR TITLE
feat: Add format "all-bytes" to archive processor

### PR DESCRIPTION
**Pull request by stergiotis**
This PR introduces the `all-bytes` codec to the `archive` processor. It has the same semantic as the `all-bytes` codec used e.g. in the `file` input and output.

I found it to be especially useful in conjecture with the `http_client`.

### CODE REVIEW

File: lib/processor/archive.go [hunk 1]
```diff
@@ -198,6 +202,39 @@ func linesArchive(hFunc headerFunc, msg types.Message) (types.Part, error) {
...
+
+	retr := make([]byte, totalLen, totalLen)
+	pos := 0
+	for _, slice := range slices {
+		pos += copy(retr[pos:], slice)
+	}
+	return retr
+}
+func allBytesArchive(hFunc headerFunc, msg types.Message) (types.Part, error) {
+	tmpParts := make([][]byte, msg.Len())
```
Comment (Jeffail) [inline lib/processor/archive.go:225]:
```suggestion
	var buf bytes.Buffer
```

File: lib/processor/archive.go [hunk 2]
```diff
@@ -198,6 +202,39 @@ func linesArchive(hFunc headerFunc, msg types.Message) (types.Part, error) {
...
+	err := msg.Iter(func(i int, part types.Part) error {
+		tmpParts[i] = part.Get()
```
Comment (Jeffail) [inline lib/processor/archive.go:227]:
```suggestion
		buf.Write(part.Get())
```

File: lib/processor/archive.go [hunk 3]
```diff
@@ -198,6 +202,39 @@ func linesArchive(hFunc headerFunc, msg types.Message) (types.Part, error) {
...
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	newPart := msg.Get(0).Copy()
+	newPart.Set(concatenate(tmpParts))
```
Comment (Jeffail) [inline lib/processor/archive.go:234]:
```suggestion
	newPart.Set(buf.Bytes())
```

**Review by Jeffail — COMMENTED** [review]
Hey @stergiotis, I've added a suggestion for using a bytes buffer for the concat, I also think the name `concatenate` for the format itself is probably better suited.

File: lib/processor/archive.go (hunk 2 above)
Comment (stergiotis) [reply lib/processor/archive.go:227]: @Jeffail
Many thanks for looking into this. My choice has todo with my mental model of the performance of  languages with a genrational garbage collector. I assume that is less costly to allocate many fixed length small arrays for a short time than to dynamically grow a single buffer. I did not benchmark though and am very new to golang.

I therefore suggest that you choose what implementation feels more idiomatic to you and matches the existing codebase.

Comment (stergiotis) [reply lib/processor/archive.go:227]: @Jeffail Regarding the naming `all-bytes`: I found it quite supristing that throughout benthos an empty `delimiter` option is synonymous to `'\n'`. I expected it to be what is labeled `all-bytes` here. 

You're right: `concatenate` seems a nice descriptive name for this operation.

**Comment by Jeffail** [comment]
Updated, fixed and merged here: https://github.com/Jeffail/benthos/commit/2cb833e696adf42ab960c25019ca2039fbb79e72

**Comment by Jeffail** [comment]
Whoops, amended that commit and force pushed as it wiped you as an author, new commit: https://github.com/Jeffail/benthos/commit/48244633964298c9a5e177c74695c4e40501206d

Thanks @stergiotis!

**Pull request closed without merging**